### ISSUE TITLE
🐛 Bug Report: Azure AD users and groups not getting ingested with the use of filter

### ISSUE THREAD

**Issue by mustafazeya** [backstage/backstage#16259]
### 📜 Description

After following the [doc](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend-module-msgraph/README.md) to use MicrosoftGraphOrgEntityProvider in backstage, Azure AD data not getting ingested

Logs shows that the data is being read `Reading msgraph users and groups` but nothing appears in Backstage portal. There is no timeout as such in Logs.

If we remove the filter all the users and groups gets ingested creating unnecessary huge refresh during every time task runs and new users and groups doesn't get ingested on time.

### 👍 Expected behavior

Users and groups should show as per the filter.

### 👎 Actual Behavior with Screenshots

No user or group getting ingested when `advanced` filter is applied.

### 👟 Reproduction steps

`App-config.production.yaml`

```yaml
catalog:
  providers:
    microsoftGraphOrg:
      production:
        target: https://graph.microsoft.com/v1.0
        authority: https://login.microsoftonline.com
        tenantId: ${AUTH_MICROSOFT_TENANT_ID}
        clientId: ${AUTH_MICROSOFT_CLIENT_ID}
        clientSecret: ${AUTH_MICROSOFT_CLIENT_SECRET}
        queryMode: advanced
        user:
          filter: endsWith(mail,'@abc.se') OR endsWith(mail,'@abc.dk') OR endsWith(mail,'@abc.no') OR endsWith(mail,'@abc.fi') OR endsWith(mail,'@abc.com')
        group:
          filter: securityEnabled eq true
        schedule:
          frequency: { minutes: 90 }
          timeout: { minutes: 80 }
          initialDelay: { seconds: 15}
       
```
     
  `catalog.ts`
  
  ```ts

import { CatalogBuilder } from '@backstage/plugin-catalog-backend';
import { ScaffolderEntitiesProcessor } from '@backstage/plugin-scaffolder-backend';
import { Router } from 'express';
import { PluginEnvironment } from '../types';
import {
  GithubDiscoveryProcessor,
  GithubOrgReaderProcessor,
} from '@backstage/plugin-catalog-backend-module-github';
import {
  ScmIntegrations,
  DefaultGithubCredentialsProvider,
} from '@backstage/integration';
import { MicrosoftGraphOrgEntityProvider } from '@backstage/plugin-catalog-backend-module-msgraph';

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  const builder = await CatalogBuilder.create(env);
  builder.setProcessingIntervalSeconds(300);
  const integrations = ScmIntegrations.fromConfig(env.config);
  const githubCredentialsProvider =
    DefaultGithubCredentialsProvider.fromIntegrations(integrations);
  // The target parameter below needs to match one of the providers' target
  // value specified in your app-config (see above).
  builder.addEntityProvider(
    MicrosoftGraphOrgEntityProvider.fromConfig(env.config, {
      logger: env.logger,
      schedule: env.scheduler.createScheduledTaskRunner({
        frequency: { minutes: 90 },
        timeout:  { minutes: 80 },
        initialDelay: { seconds: 15}
      }),
    }),
  );
  builder.addProcessor(
    GithubDiscoveryProcessor.fromConfig(env.config, {
      logger: env.logger,
      githubCredentialsProvider,
    }),
    GithubOrgReaderProcessor.fromConfig(env.config, {
      logger: env.logger,
      githubCredentialsProvider,
    }),
  );
  builder.addProcessor(new ScaffolderEntitiesProcessor());
  const { processingEngine, router } = await builder.build();
  await processingEngine.start();
  return router;
}
```
  
  
  

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by Rugvip** [comment]
Pinging a few people who have recently contributed to this module: @lsiric, @afscrome, @pjungermann. Do any of you know what might be going on here? 🙏

**Comment by afscrome** [comment]
How many users and groups do those filters match.  Most likely this is a case of the timeout not being sufficient.  Data wont show up in the catalogue until the entire data set has been processed. 

Try with some filters that match just one or two groups.  Does that import successfully?  If so, it’s just a case of needing a longer timeout.

**Comment by mustafazeya** [comment]
@afscrome There will be atleast 12K users and 10 K groups based on the filter. When I remove the filter,all the data(which is way more than with the filter) get ingested.So,I am not sure if it is timeout issue.

**Comment by afscrome** [comment]
If you're not seeing any errors, and you're sure it's not a timeout issue, I'm not sure what else it could be.  What logs do you get?  Do you get a completion error with 0 results, or do you not get anything.  Do you get any errors?  Are there any intermediate logs showing that the loading got part way (e.g. finished users, but not groups)

```
[1] 2023-01-09T20:10:37.077Z catalog info Read 0 msgraph users and 0 msgraph groups in 173.0 seconds. Committing... type=plugin class=MicrosoftGraphOrgEntityProvider taskId=MicrosoftGraphOrgEntityProvider:default:refresh taskInstanceId=4bd2db5e-943c-440d-932a-b3cddf770e5e
```

Does your filter work as expected if you plug it in to the graph explorer?

https://developer.microsoft.com/en-us/graph/graph-explorer?request=users%3F%24filter%3DendsWith(mail%2C'%40abc.se')%2BOR%2BendsWith(mail%2C'%40abc.dk')%2BOR%2BendsWith(mail%2C'%40abc.no')%2BOR%2BendsWith(mail%2C'%40abc.fi')%2BOR%2BendsWith(mail%2C'%40abc.com')&method=GET&version=v1.0&GraphUrl=https://graph.microsoft.com

**Comment by lsiric** [comment]
agree with @afscrome : 
- test with filter using explicitly only your email and only 1 group to which you belong (to ingest 1 user and 1 group)

`If we remove the filter all the users and groups gets ingested creating unnecessary huge refresh during every time task runs and new users and groups doesn't get ingested on time.` 

- your schedule is running every `90` minutes, you can decrease that to once or twice a day. 

Additionally, we had a similar issue, where ingesting 10k+ users and 10k+ groups as well, we realised that even the ingestion does take a long time, once the ingestion is complete, the transform functions are being executed for: 

1. each organisation
2. each group
3. each user

and that might take a long time, and your logs might not reflect that, since log is executed only when this process is done.

**Comment by mustafazeya** [comment]
@afscrome @lsiric I will try to change the frequency of ingestion to give the task enough time to ingest data properly.
Do you have any example where you might be using `advanced` queryMode ? It would be nice to have a look into an `app-config` file with advanced queryMode.

**Comment by brenwhyte** [comment]
If it helps any here's my working setup:

`app-config.yaml`

```
catalog:
  providers:
    microsoftGraphOrg:
      default:
        tenantId: ${AZURE_TENANT_ID}
        userGroupMember:
          expand: manager
          filter: "displayName eq 'MY-SG-NAME"
        group:
          expand: member
          filter: >
            securityEnabled eq true
            and mailEnabled eq true
        rules:
        - allow: [User, Group]
```

`catalog.ts`

```
  builder.addEntityProvider(
    MicrosoftGraphOrgEntityProvider.fromConfig(env.config, {
      logger: env.logger,
      schedule: env.scheduler.createScheduledTaskRunner({
        frequency: { hours: 12 },
        timeout: { minutes: 15 },
        initialDelay: { seconds: 30}
      }),
    }),
  );
```

With AZURE_CLIENT_ID + AZURE_CLIENT_SECRET set via environment vars.

**Comment by padupe** [comment]
I am getting the following error message on integration:
```
MicrosoftGraphOrgEntityProvider:default refresh failed Error while reading groups/****/members/microsoft.graph.user/ from Microsoft Graph: Authorization_RequestDenied - Insufficient privileges to complete the operation
```

I already validated and the credentials are correct and with the necessary permissions.

I am getting this error after upgrading to version 1.14.0

**Comment by cdavidsc007** [comment]
> I am getting the following error message on integration:
> 
> ```
> MicrosoftGraphOrgEntityProvider:default refresh failed Error while reading groups/****/members/microsoft.graph.user/ from Microsoft Graph: Authorization_RequestDenied - Insufficient privileges to complete the operation
> ```
> 
> I already validated and the credentials are correct and with the necessary permissions.
> 
> I am getting this error after upgrading to version 1.14.0

Im having this same issue, did you fixed?

**Comment by padupe** [comment]
> > I am getting the following error message on integration:
> > ```
> > MicrosoftGraphOrgEntityProvider:default refresh failed Error while reading groups/****/members/microsoft.graph.user/ from Microsoft Graph: Authorization_RequestDenied - Insufficient privileges to complete the operation
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I already validated and the credentials are correct and with the necessary permissions.
> > I am getting this error after upgrading to version 1.14.0
> 
> Im having this same issue, did you fixed?

I needed to change the query in `app-config.yaml`:

```yaml
[...]
    microsoft:
      default:
        tenantId: ${AZUREAD_TENANTID}
        clientId: ${AZUREAD_CLIENTID}
        clientSecret: ${AZUREAD_CLIENTSECRET}
        queryMode: basic
        user:
          expand: manager
        userGroupMember:
          filter: "displayName eq 'Azure Team'"
        group:
          filter: "displayName eq 'Azure Team'"
```
